Reduce time of integration tests

tiborvass opened the pull request:
Reduce time of integration tests by dividing tests into 3 parallel runs

I have to open a new PR because force push on #39680 did not work.

Related to #39651

Commit:
Jenkinsfile: move static and cross compilation to unit-validate stage

Signed-off-by: Tibor Vass <tibor@docker.com>

Commit:
hack: unmount leftover daemon root folders

Signed-off-by: Tibor Vass <tibor@docker.com>

thaJeztah commented on hack/make/test-integration:
```diff
@@ -4,24 +4,25 @@ set -e -o pipefail
-	build_test_suite_binaries
-	bundle .integration-daemon-start
-	bundle .integration-daemon-setup
+       build_test_suite_binaries
```
This changed from tabs to spaces 😇

tiborvass replied:
blasphemy! shame on me

tiborvass commented on Jenkinsfile:
```diff
@@ -173,19 +199,46 @@ pipeline {
-                                    cross
+				rm -rf "$WORKSPACE/bundles"
+
+# run_tests() { [ -n "$TESTDEBUG" ] && rm= || rm=--rm; docker run $rm -it --privileged   -v "$WORKSPACE/bundles:/go/src/github.com/docker/docker/bundles"   -v "$WORKSPACE/.git:/go/src/github.com/docker/docker/.git"   --name "$CONTAINER_NAME"                                           -e KEEPBUNDLE=1   -e TESTDEBUG                                           -e TESTFLAGS                               ...
```
this is debug leftover, will remove

Commit:
Jenkinsfile: reduce time of integration tests by dividing tests into 3 parallel runs

Signed-off-by: Tibor Vass <tibor@docker.com>

kolyshkin commented on hack/make/.integration-daemon-stop:
```diff
@@ -10,6 +10,8 @@ if [ ! "$(go env GOOS)" = 'windows' ]; then
 			echo >&2 "warning: PID $pid from $pidFile had a nonzero exit code"
 		fi
+		root=$(dirname "$pidFile")/root
+		grep -qE "\b$root\b" /proc/mounts && umount "$root"
```
It's better to not parse /proc/mounts before calling umount; easier to ignore non-zero exit code
```
umount "$root" || true
```

kolyshkin replied:
Reason is, generating /proc/mounts is relatively expensive and sometimes error-prone so there is a chance you might not find your mount even if it's there.

andrewhsu commented on Jenkinsfile:
```diff
@@ -200,18 +200,54 @@ pipeline {
+                                          -e DOCKER_GRAPHDRIVER \
+                                          docker:${GIT_COMMIT} \
+                                          hack/make.sh \
+                                            dynbinary-daemon \
```
looks like this `dynbinary-daemon` will be called 3 times

andrewhsu replied:
having this cleaned up can be cleaned up in a follow up PR

tiborvass replied:
Done in this one

andrewhsu reviewed (approved):
LGTM w.r.t. reducing janky PR check times from 2 hrs down to 1 hr

andrewhsu committed:
add powerpc-master stage to Jenkinsfile

The powerpc-master stage will just run the integration-cli tests. The
existing powerpc stage will run the unit tests and the integration
tests. In this way, PR check jobs will be shorter, but all integration
tests will run after PR is merged to master.

Signed-off-by: Andrew Hsu <andrewhsu@docker.com>

andrewhsu committed:
add z-master stage to Jenkinsfile

The z-master stage will just run the integration-cli tests. The
existing z stage will run the unit tests and the integration
tests. In this way, PR check jobs will be shorter, but all
integration tests will run after PR is merged to master.

Signed-off-by: Andrew Hsu <andrewhsu@docker.com>

andrewhsu committed:
set timeouts in Jenkinsfile to 2 hrs

Signed-off-by: Andrew Hsu <andrewhsu@docker.com>

andrewhsu commented:
With the last 4 commits I added, the total time for the PR check with `Jenkinsfile` takes 59m 30s:
https://ci.docker.com/public/blue/organizations/jenkins/moby/detail/PR-39682/25/pipeline/

This was accomplished by moving the integration-cli tests to only run on merge to master for experimental, z, and powerpc builds. The integration-cli tests are still run on PR checks in the janky job. The integration-cli tests will likely add 1 hr of time to the jobs that run against master branch, so it will be around 2 hrs total: https://ci.docker.com/public/blue/organizations/jenkins/moby/activity?branch=master

thaJeztah commented:
w.r.t. the new `experimental-master`; https://github.com/moby/moby/pull/39709

thaJeztah commented on Jenkinsfile:
```diff
@@ -283,6 +351,76 @@ pipeline {
                         }
                     }
                 }
+                stage('experimental-master') {
```
Opened https://github.com/moby/moby/pull/39711 to try this approach

andrewhsu replied:
Thanks for trying out an alternative approach, but the reason to have a separate stage with its own agent for running on master would be to keep the overall time under 2 hrs. Granted, that means we will have a job run on the master branch use two z machines instead of just one z machine for a PR job, but it will keep the master job running under 1 hour in total.

If we run the integration and integration-cli tests on the same node for z machines, it will sometimes take over 2 hrs so we will then have to change the timeout settings to be different for a PR and for a master job.

The different filenames for `z-bundles.tar.gz` and `z-master-bundles.tar.gz` is to help a user distinguish the difference between the two if they download both from the same artifacts page of the jenkins job.

I agree there is a bit of duplicate code going on in the `Jenkinsfile` because of this, but I think we could simplifying into Makefile targets which is what a human would want anyways to easily reproduce a failure observed in a jenkins job.

For now, concentrating on getting jenkins jobs below 1 hour as top priority. The 1 hour we save now will help our future PRs to make the code cleaner.

thaJeztah replied:
Ah, gotcha; it's to run them in parallel. 

Perhaps rename them to what they actually do, instead of the branch they're running on? So `XX-integration`, `XX-integration-cli`

We should try to use the same approach as is done on Janky (if possible), but current Power and Z machines have less memory

This https://github.com/moby/moby/pull/39708 (or similar) would also help reduce the duplication in code, and keep the Jenkinsfile a bit more manageable

andrewhsu commented on Jenkinsfile:
```diff
@@ -4,7 +4,7 @@ pipeline {
     options {
         buildDiscarder(logRotator(daysToKeepStr: '30'))
-        timeout(time: 3, unit: 'HOURS')
+        timeout(time: 2, unit: 'HOURS')
```
The global timeout is for jenkins to keep track of the time. The `TIMEOUT` for integration tests are for the tests that honor that environment variable to keep track of its own time and abort if it runs over regardless if the node loses connection with jenkins.

I think for simplicity of maintenance, it is just easier to set everything to the same 2-hr timeout value.

thaJeztah replied:
Yeah, shouldn't be a big issue; My train of thought was that 2hrs (max; integration) + other steps > 2hrs

andrewhsu commented on Jenkinsfile:
```diff
@@ -414,7 +626,8 @@ pipeline {
                                   -e DOCKER_GRAPHDRIVER \
-                                  -e TIMEOUT="180m" \
+                                  -e TEST_SKIP_INTEGRATION_CLI=1 \
+                                  -e TIMEOUT='120m' \
```
sounds good. i'll do this.

andrewhsu commented on Jenkinsfile:
```diff
@@ -414,7 +626,8 @@ pipeline {
                                   -e DOCKER_GITCOMMIT=${GIT_COMMIT} \
                                   -e DOCKER_GRAPHDRIVER \
-                                  -e TIMEOUT="180m" \
+                                  -e TEST_SKIP_INTEGRATION_CLI=1 \
```
i'll make these changes.

andrewhsu commented on Jenkinsfile:
```diff
@@ -78,6 +78,33 @@ pipeline {
                                 '''
                             }
                         }
+			stage("Static") {
```
done

thaJeztah replied:
looks like you did it in the wrong commit 😅 😇

andrewhsu committed:
use environment for power jobs in Jenkinsfile

Signed-off-by: Andrew Hsu <andrewhsu@docker.com>

andrewhsu committed:
use environment for z jobs in Jenkinsfile

Signed-off-by: Andrew Hsu <andrewhsu@docker.com>

thaJeztah commented on Jenkinsfile:
```diff
@@ -307,6 +377,80 @@ pipeline {
                         }
                     }
                 }
+                stage('z-master') {
```
left that as an earlier comment (but perhaps it got buried; consider renaming these to;

- `z-integration`
- `z-integration-cli`

(and rename the bundles accordingly); it describes better what they are, and make it easier to find the right `bundle.tar.gz` if you're looking for a specific test failure

andrewhsu replied:
done

andrewhsu committed:
be more lenient on junit report gathering in Jenkinsfile

In case a job fails before even generating a report file.

Signed-off-by: Andrew Hsu <andrewhsu@docker.com>

andrewhsu committed:
rename z bundles in Jenkinsfile

Signed-off-by: Andrew Hsu <andrewhsu@docker.com>

andrewhsu committed:
rename powerpc bundles in Jenkinsfile

Signed-off-by: Andrew Hsu <andrewhsu@docker.com>

andrewhsu commented:
Hmm...looks like the docker-py tests are [failing and hanging](https://ci.docker.com/public/job/moby/job/PR-39682/32/execution/node/173/log/) with the error:
```
=================================== FAILURES ===================================
 [31m[1m___________________ PullImageTest.test_pull_invalid_platform ___________________[0m
 [1m[31mtests/integration/api_image_test.py[0m:72: in test_pull_invalid_platform
 [1m    assert excinfo.value.status_code == 500[0m
 [1m[31mE   AssertionError: assert 400 == 500[0m
 [1m[31mE    +  where 400 = APIError(HTTPError('400 Client Error: Bad Request for url: http+docker://localhost/v1.35/images/create?fromImage=hello-world&platform=foobar',),).status_code[0m
 [1m[31mE    +    where APIError(HTTPError('400 Client Error: Bad Request for url: http+docker://localhost/v1.35/images/create?fromImage=hello-world&platform=foobar',),) = <ExceptionInfo APIError tblen=4>.value[0m
```

thaJeztah commented:
That's weird; I think I excluded that test because it was broken

thaJeztah commented on hack/make/test-docker-py:
```diff
@@ -17,6 +17,7 @@ source hack/make/.integration-test-helpers
 --deselect=tests/integration/api_exec_test.py::ExecDemuxTest::test_exec_command_tty_stream_no_demux \
 --deselect=tests/integration/api_build_test.py::BuildTest::test_build_invalid_platform \
 --deselect=tests/integration/api_image_test.py::PullImageTest::test_build_invalid_platform \
+--deselect=tests/integration/api_image_test.py::PullImageTest::test_pull_invalid_platform \
```
ah! thx!

thaJeztah replied:
I think the one above may be incorrect then

thaJeztah replied:
opened as a separate PR so that we get master to go green again while this is still being reviewed https://github.com/moby/moby/pull/39714

thaJeztah commented on Jenkinsfile:
```diff
@@ -65,21 +66,50 @@ pipeline {
                         }
                         stage("Docker-py") {
+                            steps {
+                                ansiColor('xterm') {
```
I noticed the output was _sometimes_ wrong, and _sometimes_ looked good; I initially disabled color output with `--color=no` on `pytest`, but we actually didn't need a TTY for these tests, so removing the `--t` did the right thing (https://github.com/moby/moby/pull/39716)

andrewhsu replied:
i've removed the `ansiColor` commit

andrewhsu committed:
fix bundles filenames in Jenkinsfile

Signed-off-by: Andrew Hsu <andrewhsu@docker.com>

andrewhsu commented:
I've rebased to resolve conflicts.

With these final changes in, would like to merge when green so we can reap the benefits of faster PR check results.

andrewhsu commented:
From the PR check on eb30f0a there is just a [known](https://github.com/moby/moby/issues/39499) flaky test [failure](https://ci.docker.com/public/blue/organizations/jenkins/moby/detail/PR-39682/39/pipeline):
```
FAIL: docker_cli_swarm_test.go:1333: DockerSwarmSuite.TestSwarmClusterRotateUnlockKey
```

The legacy PR check on janky has the [same test passing](https://jenkins.dockerproject.org/job/Docker-PRs/55421/console).
```
PASS: docker_cli_swarm_test.go:1333: DockerSwarmSuite.TestSwarmClusterRotateUnlockKey	70.303s
```

andrewhsu commented:
@thaJeztah @vdemeester @cpuguy83 this PR is ready for your consideration

thaJeztah commented on Jenkinsfile:
```diff
@@ -80,6 +81,33 @@ pipeline {
+                                '''
+                            }
+                        }
+                        // needs to be last stage that calls make.sh for the junit report to work
```
What was the problem if it wasn't the last stage? Does it help if you upload the junit.xml as part of the stage itself? (See https://github.com/moby/moby/pull/39719)

andrewhsu replied:
The unit test stage needs to be last because the tests will be written to `bundles` directory and if the other stages run after the unit test stage then the directory would be wiped out.

Want to honor the return value of the unit tests, so if the tests has a failure then it will return non-zero back to the shell and jenkins will stop processing that step and mark that step as red. The `always` stage is an exception catch stage that will be run whether any of the steps fail, so that would be a suitable place to consume the junit.xml file.

cpuguy83 replied:
Does KEEPBUNDLE=1 work?

tiborvass replied:
Maybe. Can do in followup

thaJeztah replied:
Each stage can upload its bundles/results before the next stage starts

thaJeztah reviewed (approved):
LGTM

tiborvass commented:
LGTM

cpuguy83 reviewed (approved):
LGTM

It would be nice if we can consolidate the actual logic and only setup the envs for each arch.

thaJeztah commented:
@cpuguy83 yes; was also giving things a quick go in https://github.com/moby/moby/pull/39708
